test/testdata/infer/metatype_new.rb

```diff
@@ -0,0 +1,37 @@
+# typed: true
+
+class MyGeneric
+  extend T::Generic
+
+  X = type_member
+end
+
+MyArrayString = T.type_alias {T::Array[String]}
+
+# -- good --
+
+String.new
+T::Array[String].new
+MyGeneric.new
+MyGeneric[Integer].new
+
+# -- bad, but we can't detect them yet --
+
+MyArrayString.new                  # Type alias is unwound by the type we get to MetaType::dispatchCall
+
+# -- bad --
+
+T.any(Integer, String).new         # error: Cannot call `new` on type `T.any(Integer, String)`
+T.all(Kernel, Comparable).new      # error: Cannot call `new` on type `T.all(Kernel, Comparable)`
+T.nilable(String).new              # error: Cannot call `new` on type `T.nilable(String)`
+
+T.untyped.new                      # error: Cannot call `new` on type `T.untyped`
+T.self_type.new                    # error: Cannot call `new` on type `T.untyped`
+
+T.class_of(String).new             # error: Cannot call `new` on type `T.class_of(String)`
+T.class_of(MyGeneric).new          # error: Cannot call `new` on type `T.class_of(MyGeneric)`
+T.class_of(MyGeneric[Integer]).new # error: Cannot call `new` on type `T.class_of(MyGeneric)`
```

Comment:
I sure hope we don't even parse `T. class_of(MyGeneric[Integer]).new`. Does this test pass? I would have hoped there was another error reported on this line in addition to the `Cannot call new` error.

Reply:
The test does pass, yeah. :)

Seems that we do parse it:

https://sorbet.run/#class%20MyGeneric%0A%20%20extend%20T%3A%3AGeneric%0A%0A%20%20X%20%3D%20type_member%0Aend%0A%0AT.class_of(MyGeneric%5BInteger%5D).new

Reply:
looks like we only report the error when it's for sure in type syntax position

https://sorbet.run/#%23%20typed%3A%20true%0Aextend%20T%3A%3ASig%0A%0Asig%20%7Bparams(x%3A%20T.class_of(T%3A%3AArray%5BInteger%5D)).void%7D%0Adef%20foo(x)%0A%20%20T.reveal_type(x)%0Aend%0A%0AT%3A%3AArray%5BT.class_of(T%3A%3AArray%5BInteger%5D)%5D.new

We should probably have a separate ticket to make these two cases consistent.

Reply:
#4377